### ISSUE TITLE
Single match arm not causing `single_match` warning

### ISSUE THREAD

**Issue by bingcicle** [rust-lang/rust-clippy#8969]
### Summary

Was writing code as seen below, ran `cargo clippy` with no warnings. (note: this is just a short version of what I actually wrote)

Is there a reason clippy is lenient with this particular check not warning for `single_match`, or is this simply the case of an uncaught lint?

### Reproducer

I tried this code:

```rust
use std::io::Error;

fn parse() -> Result<(), Error> {
    Ok(())
}

fn main() {
    match parse() {
        Ok(_) => {}
        Err(e) => println!("Error: {}", e),
    }
}
```

I expected to see the `single_match` lint warning ([Clippy Lints page](https://rust-lang.github.io/rust-clippy/master/index.html#single_match)) here, but instead `cargo clippy` runs successfully without complaint.

 Took a second look and realized that the code could be re-written with the equivalent below:

```rust
...
fn main() {
    if let Err(e) = parse() {
        println!("Error: {}", e)
    }
}
```

### Version

```text
rustc 1.61.0 (fe5b13d68 2022-05-18)
binary: rustc
commit-hash: fe5b13d681f25ee6474be29d748c65adcd91f69e
commit-date: 2022-05-18
host: x86_64-apple-darwin
release: 1.61.0
LLVM version: 14.0.0
```

**Comment by tamaroning** [comment]
This is not a bug.
Please see https://github.com/rust-lang/rust-clippy/issues/8282

**Comment by jonathanberthias** [comment]
I think this case is a little more complicated, if we change the order of the matches then the lint fires:
```rust
use std::io::Error;

fn parse() -> Result<(), Error> {
    Ok(())
}

fn main() {
    match parse() {
        Err(e) => println!("Error: {}", e),
        Ok(_) => {}
    }
}
```

Clippy will output the expected:
```
warning: you seem to be trying to use `match` for destructuring a single pattern. Consider using `if let`
   --> tests/integration_tests.rs:335:5
    |
335 | /     match parse() {
336 | |         Err(e) => println!("Error: {}", e),
337 | |         Ok(_) => {}
338 | | }
    | |_^ help: try: `if let Err(e) = parse() { println!("Error: {}", e) }`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#single_match
```

I assume order shouldn't matter for this rule, which means this is a genuine bug, no?
